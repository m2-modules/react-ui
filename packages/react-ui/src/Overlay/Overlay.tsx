import { useCallback, useEffect, useRef, useState } from "react";
import { createPortal } from "react-dom";
import usePortal from "../hooks/usePortal";

export interface OverlayProps {
  active?: boolean;
  activateTransitionClasses?: string[];
  inactivateTransitionClasses?: string[];
  transitionDuration?: string;
  background?: string;
  portalIdentifier?: string;
  handleClick?: () => void;
}

export default function Overlay({
  active = false,
  activateTransitionClasses = ["opacity-1"],
  inactivateTransitionClasses = ["opacity-0"],
  transitionDuration = "duration-500",
  background = "rgba(0,0,0,0.3)",
  portalIdentifier = "overlay-portal",
  handleClick,
}: OverlayProps) {
  const overlayRef = useRef<HTMLDivElement>(null);
  const overlayPortal = usePortal(portalIdentifier);
  const [_active, _setActive] = useState(active);

  const activateOverlay = useCallback(() => {
    if (!overlayRef.current) return;
    const overlay = overlayRef.current;
    _setActive(true);

    setTimeout(() => {
      overlay.classList.remove(...inactivateTransitionClasses);
      overlay.classList.add(...activateTransitionClasses);
    }, 100);
  }, [activateTransitionClasses, inactivateTransitionClasses]);

  const inactivateOverlay = useCallback(() => {
    if (!overlayRef.current) return;
    const overlay = overlayRef.current;
    overlay.classList.remove(...activateTransitionClasses);
    overlay.classList.add(...inactivateTransitionClasses);
  }, [activateTransitionClasses, inactivateTransitionClasses]);

  useEffect(() => {
    if (active) {
      activateOverlay();
    } else {
      inactivateOverlay();
    }
  }, [active, activateOverlay, inactivateOverlay]);

  if ((!active && !_active) || !overlayPortal) return <></>;

  return createPortal(
    <div
      ref={overlayRef}
      className={`fixed left-0 right-0 top-0 bottom-0 z-[2] transition-all ${transitionDuration} ${inactivateTransitionClasses.join(
        " ",
      )}`}
      style={{ background }}
      onTransitionEnd={() => {
        if (!active && _active) _setActive(false);
      }}
      onClick={handleClick}
    />,
    overlayPortal,
  );
}
