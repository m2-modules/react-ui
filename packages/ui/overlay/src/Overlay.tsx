import React from "react";
import { useCallback, useEffect, useRef } from "react";
import { createPortal } from "react-dom";

const ACTIVE_CLASS = "opacity-1";
const INACTIVE_CLASS = "opacity-0";

const getOverlayPortal = () => {
  let overlayPortal = document.querySelector<HTMLDivElement>("#overlay-portal");
  if (overlayPortal) return overlayPortal;
  overlayPortal = document.createElement("div");
  overlayPortal.id = "overlay-portal";
  document.body.append(overlayPortal);
  return overlayPortal;
};
export interface OverlayProps {
  active?: boolean;
  background?: string;
  onClose: () => void;
}

export default function Overlay({
  active = false,
  background = "rgba(0,0,0,0.3)",
  onClose,
}: OverlayProps) {
  const overlayRef = useRef<HTMLDivElement>(null);

  const activateOverlay = useCallback(() => {
    if (!overlayRef.current) return;
    const overlay = overlayRef.current;

    setTimeout(() => {
      overlay.classList.remove(INACTIVE_CLASS);
      overlay.classList.add(ACTIVE_CLASS);
    }, 100);
  }, []);

  const inactivateOverlay = useCallback(() => {
    if (!overlayRef.current) return;
    const overlay = overlayRef.current;
    overlay.classList.remove(ACTIVE_CLASS);
    overlay.classList.add(INACTIVE_CLASS);
  }, []);

  useEffect(() => {
    if (active) {
      activateOverlay();
    } else {
      inactivateOverlay();
    }
  }, [active, activateOverlay, inactivateOverlay]);

  return createPortal(
    <div
      ref={overlayRef}
      className="fixed left-0 right-0 top-0 bottom-0 transition-opacity duration-500 opacity-0"
      style={{ background }}
      onClick={onClose}
    />,
    getOverlayPortal(),
  );
}
