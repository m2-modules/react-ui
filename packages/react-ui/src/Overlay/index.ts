import Overlay from "./Overlay";

export default Overlay;
export * from "./Overlay";
