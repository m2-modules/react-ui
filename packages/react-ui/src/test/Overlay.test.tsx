import renderer from "react-test-renderer";
import Overlay from "../Overlay";

it("show Overlay on document", () => {
  const component = renderer.create(
    <Overlay
      handleClick={() => {
        console.log("On close");
      }}
    />,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
