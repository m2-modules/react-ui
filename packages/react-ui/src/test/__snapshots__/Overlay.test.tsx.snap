// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show Overlay on document 1`] = `null`;
